pygmt/src/grd2xyz.py

```diff
@@ -0,0 +1,78 @@
+"""
+grd2xyz - Convert grid to data table
+"""
+import numpy as np
+import pandas as pd
+from pygmt.clib import Session
+from pygmt.exceptions import GMTInvalidInput
+from pygmt.helpers import (
+    GMTTempFile,
+    build_arg_string,
+    fmt_docstring,
+    kwargs_to_strings,
+    use_alias,
+)
+
+
+@fmt_docstring
+@use_alias(
+    R="region",
+    V="verbose",
+)
+@kwargs_to_strings(R="sequence")
+def grd2xyz(grid, output_type="d", **kwargs):
+    r"""
+    Create xyz tables from grid files.
+
+    Read a binary 2-D grid files and write out
+    xyz-triplets in ASCII [or binary] format to a standard output.
+
+    Full option list at :gmt-docs:`grd2xyz.html`
+
+    {aliases}
+
+    Parameters
+    ----------
+    grid : str or xarray.DataArray
+        The file name of the input grid or the grid loaded as a DataArray.
+        This is the only required parameter.
+    output_type : str
+        Determine the format the xyz data will be returned in:
+            **a**: numpy array
+            **d**: pandas DataFrame [Default option]
+            **s**: string
+
+    {R}
+    {V}
+
+    Returns
+    -------
+    info : pandas.DataFrame or numpy.array or str
+        The xyz triplet data in a pandas DataFrame, numpy array, or string.
+    """
+    with GMTTempFile() as outfile:
+        with Session() as lib:
+            file_context = lib.virtualfile_from_data(check_kind="raster", data=grid)
+            with file_context as infile:
+                arg_str = " ".join(
+                    [infile, build_arg_string(kwargs), "->" + outfile.name]
+                )
+                lib.call_module("grd2xyz", arg_str)
+        result = outfile.read()
+    if output_type == "s":
+        return result
+    data_list = []
+    for string_entry in result.strip().split("\n"):
+        float_entry = []
+        string_list = string_entry.strip().split()
+        for i in string_list:
+            float_entry.append(float(i))
+        data_list.append(float_entry)
+    data_array = np.array(data_list)
+    if output_type == "a":
+        result = data_array
+    elif output_type == "d":
+        result = pd.DataFrame(data_array)
+    else:
+        raise GMTInvalidInput("""Must specify format as either a, d, or s.""")
```

Comment:
```suggestion
    if output_type not in ["a", "d", "s"]:
        raise GMTInvalidInput("""Must specify format as either a, d, or s.""")
    if output_type == "s" and outfile is None:
        raise GMTInvalidInput("""Must specify outfile for ASCII output.""")

    with GMTTempFile() as tmpfile:
        with Session() as lib:
            file_context = lib.virtualfile_from_data(check_kind="raster", data=grid)
            with file_context as infile:
                if outfile is None:
                    outfile = tmpfile.name
                arg_str = " ".join([infile, build_arg_string(kwargs), "->" + outfile])
                lib.call_module("grd2xyz", arg_str)

        # Read temporary csv output to a pandas table
        if outfile == tmpfile.name:  # if user did not set outfile, return pd.DataFrame
            result = pd.read_csv(tmpfile.name, sep="\t", header=None, comment=">")
        elif outfile != tmpfile.name:  # return None if outfile set, output in outfile
            result = None

        if output_type == "a":
            result = result.to_numpy()
```

I tested this implementation (which is based on the blockmean code), which supports ASCII file output when output_type is "s" and uses pandas.read_csv to get a DataFrame output rather than outfile.read() and string formatting. It's a lot faster for numpy and pandas outputs (~130ms versus ~180ms on average) for running grd2xyz a 240 x 240 grid. I also think it's a bit more intuitive for maintenance. Using this implementation would require updating /pygmt/tests/test_grd2xyz.py::test_grd2xyz and removing the numpy import.